main.rb

```diff
@@ -0,0 +1,208 @@
+          put_req
+        else
+          raise "Unsupported HTTP method: #{method}"
+        end
+
+  Net::HTTP.start(uri.hostname, uri.port, use_ssl: $use_ssl) { |http| http.request(req) }
+end
+
+def handle_api_response(res, action, parsed = true)
+  case res
+  when Net::HTTPSuccess
+    puts "#{action.capitalize} successful.".green
+    return parsed ? get_parsed_response(res.body) : nil
+  when Net::HTTPUnauthorized
+    puts "Authorization error while #{action}: #{get_parsed_response(res.body)[:error_description]}".red
+  when Net::HTTPClientError
+    puts "Client error while #{action}: #{get_parsed_response(res.body)[:message]}".red
+  else
+    puts "Unexpected error while #{action}: #{get_parsed_response(res.body)}".red
+  end
+  return nil
+end
+
+def login
+  puts "Logging in to Testinium...".yellow
+  base_url = $ent_base_url ? get_base_url.sub("/api", "") : "https://account.testinium.com/uaa"
+
+  # Testinium's login API uses a public generic token for authentication. More details:  
+  # Cloud: https://testinium.gitbook.io/testinium/apis/auth/login  
+  # Enterprise: https://testinium.gitbook.io/testinium-enterprise/apis/auth/login 
+  token = $ent_base_url ? "Y2xpZW50MTpjbGllbnQx" : "dGVzdGluaXVtU3VpdGVUcnVzdGVkQ2xpZW50OnRlc3Rpbml1bVN1aXRlU2VjcmV0S2V5"
+  url = "#{base_url}/oauth/token"
+  headers = { 'Content-Type' => 'application/x-www-form-urlencoded', 'Authorization' => "Basic #{token}" }
+  body = { 'grant_type' => 'password', 'username' => $username, 'password' => $password }
+
+  retry_request($each_api_max_retry_count) do
+    res = send_request('POST', url, headers, body)
+    parsed_response = handle_api_response(res, "logging")
+    return parsed_response[:access_token] if parsed_response
+  end
+end
+
+def find_project(access_token)
+  puts "Searching for project #{$project_id}...".blue
+  url = "#{get_base_url}/projects/#{$project_id}"
+  headers = { 'Content-Type' => 'application/json', 'Authorization' => "Bearer #{access_token}", 'current-company-id' => $company_id }
+
+  retry_request($each_api_max_retry_count) do
+    res = send_request('GET', url, headers)
+    parsed_response = handle_api_response(res, "searching the project")
+    return parsed_response if parsed_response
+  end
+end
+
+def upload(access_token)
+  puts "Uploading #{$file_name} to Testinium...".yellow
+  url = $ent_base_url ? "#{get_base_url}/file/project/#{$project_id}" : "#{get_base_url}/file/upload"
+  file_form = $ent_base_url ? "files" : "file"
+  headers = { 'Authorization' => "Bearer #{access_token}", 'current-company-id' => $company_id }
+  form_data = { file_form => File.open($file), 'isSignRequired' => 'true' }
+
+  retry_request($each_api_max_retry_count) do |count|
+    puts "File upload attempt: #{count}".blue
+    res = send_request('POST', url, headers, form_data)
+    parsed_response = handle_api_response(res, "uploading the application", !$ent_base_url)
+    return parsed_response
```

Comment:
In the handle_api function, we try to catch errors, but we should handle here like if parsed_reponse is available return result

```suggestion
    return parsed_response if parsed_response
```